window_identifier: Add `set_parent_of` method

arun-mani-j opened the pull request:
This is commonly required in backend implementations as the portal dialog has to be set as modal with respect to the original application that launched them.

Not sure if I understand the macros control flow properly, any suggestions on improvement is very much welcome.

bilelmoussaoui commented on src/window_identifier/mod.rs:
```diff
@@ -310,6 +314,47 @@ impl Serialize for WindowIdentifierType {
+
+        let surface = gtk::prelude::NativeExt::surface(window).unwrap();
+
+        match self {
+            #[cfg(feature = "gtk4_x11")]
+            WindowIdentifierType::X11(_) => {
+                // TODO
+                false
+            }
+            #[cfg(feature = "gtk4_wayland")]
+            WindowIdentifierType::Wayland(handle) => {
+                let toplevel =
+                    gtk::glib::object::Cast::dynamic_cast::<WaylandToplevel>(surface).unwrap();
+                toplevel.set_transient_for_exported(handle)
+            }
+            _ => false,
```
that is expected yes. You would have to do a `#[cfg(not(all(feature = "gtk4_x11", feature = "gtk4_wayland")))]` or so to avoid the clippy lint above

arun-mani-j replied:
Thanks, done.

bilelmoussaoui commented on src/window_identifier/mod.rs:
```diff
@@ -310,6 +314,47 @@ impl Serialize for WindowIdentifierType {
+    ///
+    /// Returns `true` on success.
+    #[allow(unused_variables)]
+    #[cfg(any(feature = "gtk4_wayland", feature = "gtk4_x11"))]
```
```suggestion
    #[cfg(all(any(feature = "gtk4_wayland", feature = "gtk4_x11")), feature = "backend")]
```
to avoid it being exposed all the time

bilelmoussaoui commented on src/window_identifier/mod.rs:
```diff
@@ -310,6 +314,47 @@ impl Serialize for WindowIdentifierType {
+    #[allow(unused_variables)]
+    #[cfg(any(feature = "gtk4_wayland", feature = "gtk4_x11"))]
+    pub fn set_parent_of(&self, window: &gtk::Window) -> bool {
+        let fake_parent = gtk::Window::new();
+        gtk::prelude::GtkWindowExt::set_transient_for(window, Some(&fake_parent));
+        gtk::prelude::GtkWindowExt::set_modal(window, true);
+
+        if gtk::prelude::NativeExt::surface(window).is_none() {
+            gtk::prelude::NativeExt::realize(window);
+        }
+
+        let surface = gtk::prelude::NativeExt::surface(window).unwrap();
```
```suggestion
        use gtk::prelude::*;
        let fake_parent = gtk::Window::new();
        window.set_transient_for(Some(&fake_parent));
        window.set_modal(true);

        if window.surface().is_none() {
            window.realize();
        }

        let surface = window.surface().unwrap();
```

bilelmoussaoui commented on src/window_identifier/mod.rs:
```diff
@@ -310,6 +314,47 @@ impl Serialize for WindowIdentifierType {
 }
 
+impl WindowIdentifierType {
+    /// Sets the given window as a modal child of the window represented by
+    /// self. Currently only Wayland windows are supported.
```
You should also add a note that this will only work if a wayland window is placed above another wayland window, not above an x11 one. For that one would need a private wayland protocol from the compositor to do so, mutter does that with `x11_interop` and https://gitlab.gnome.org/GNOME/libgxdp/-/tree/main?ref_type=heads

bilelmoussaoui commented on src/window_identifier/mod.rs:
```diff
@@ -310,6 +314,47 @@ impl Serialize for WindowIdentifierType {
+
+        match self {
+            #[cfg(feature = "gtk4_x11")]
+            WindowIdentifierType::X11(_) => {
+                // TODO
+                false
```
This https://github.com/flatpak/xdg-desktop-portal-gtk/blob/main/src/externalwindow-x11.c#L48-L110 can help for implementing the X11 side

arun-mani-j replied:
Thanks done.

So:
- Wayland application -> Wayland server = works fine.
- Wayland application -> X11 server = ends in warning as expected: can not cast X11Toplevel to WaylandToplevel.
- X11 application -> X11 server = GTK says it is unable to find the surface corresponding to the XID, I don't understand what I'm doing wrong :thinking:. 
- X11 application -> Wayland server = ends in warning as expected (I think?): can not find surface corresponding to XID.

If we fix the X11 application to X11 server functionality, then it is all set I think.

bilelmoussaoui replied:
>Wayland application -> X11 server = ends in warning as expected: can not cast X11Toplevel to WaylandToplevel.
Can this even happen?

arun-mani-j replied:
Yep, launch the server under X11 (`GDK_BACKEND=x11`) and application under Wayland. 

So now the portal dialog is of X11Toplevel type and so can not be cast to a WaylandToplevel. So it fails with warning here - https://github.com/bilelmoussaoui/ashpd/pull/288/files#diff-54d5d9106ea94967f399a2c21ee27948ae23aa5b39116a93bdabbefb691ece18R394

arun-mani-j commented:
v1:
- Cleaned up the code a bit, thanks @bilelmoussaoui.
- Added functionality to set X11 window as parent for another X11 window.
- Removed `unwrap` on objects that are based on external objects; we don't want the server to crash just because a rogue client sent an invalid XID.

bilelmoussaoui commented on src/window_identifier/mod.rs:
```diff
@@ -310,6 +315,94 @@ impl Serialize for WindowIdentifierType {
+            #[cfg(feature = "gtk4_x11")]
+            WindowIdentifierType::X11(xid) => {
+                use gdk4x11::{X11Display, X11Surface};
+
```
maybe it is worth trying what libgxdp is using here https://gitlab.gnome.org/GNOME/libgxdp/-/blob/main/src/gxdp-external-window-x11.c?ref_type=heads#L90-105?

bilelmoussaoui replied:
can you add a comment pointing to where you "copied" the implementation from?

arun-mani-j replied:
Done.

arun-mani-j commented on src/window_identifier/mod.rs:
```diff
@@ -310,6 +315,94 @@ impl Serialize for WindowIdentifierType {
+                };
+                let foreign_window = match native.dynamic_cast::<gtk::Window>() {
+                    Ok(window) => window,
+                    Err(_) => {
```
You mean catch the error in some variable instead of `_` and log it? From the docs, `dynamic_cast` returns `Result<T, Self>`, so there is no error message right? :thinking:

bilelmoussaoui replied:
ah my bad then

arun-mani-j commented on src/window_identifier/mod.rs:
```diff
@@ -310,6 +315,94 @@ impl Serialize for WindowIdentifierType {
+        any(feature = "gtk4_wayland", feature = "gtk4_x11"),
+        feature = "backend"
+    ))]
+    pub fn set_parent_of(&self, window: &gtk::Window) -> bool {
```
I'm not sure if I understand how `IsA` works correctly, but if we are changing to it, it seems we need to do a lot of casts? Because I'm getting these compilation errors by changing to `IsA`:

```
❯ cargo build --features backend,gtk4_wayland,gtk4_x11,tracing
   Compiling ashpd v0.11.0 (/home/arun-mani-j/Projects/Phosh/ashpd)
error[E0599]: the method `surface` exists for reference `&impl gtk::prelude::IsA<gtk::Window>`, but its trait bounds were not satisfied
   --> src/window_identifier/mod.rs:337:36
    |
337 |         let surface = match window.surface() {
    |                                    ^^^^^^^ method cannot be called on `&impl gtk::prelude::IsA<gtk::Window>` due to unsatisfied trait bounds
    |
    = note: the following trait bounds were not satisfied:
            `impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<Drag>`
            which is required by `impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::DragExt`
            `impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<DrawContext>`
            which is required by `impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::DrawContextExt`
            `impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<GLContext>`
            which is required by `impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::GLContextExt`
            `impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<gtk4::gsk4::Renderer>`
            which is required by `impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::GskRendererExt`
            `impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<Native>`
            which is required by `impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::NativeExt`
            `&impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<Drag>`
            which is required by `&impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::DragExt`
            `&impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<DrawContext>`
            which is required by `&impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::DrawContextExt`
            `&impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<GLContext>`
            which is required by `&impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::GLContextExt`
            `&impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<gtk4::gsk4::Renderer>`
            which is required by `&impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::GskRendererExt`
            `&impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<Native>`
            which is required by `&impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::NativeExt`
    = help: items from traits can only be used if the type parameter is bounded by the trait
help: the following traits define an item `surface`, perhaps you need to restrict type parameter `impl gtk::prelude::IsA<gtk::Window>` with one of them:
    |
334 |     pub fn set_parent_of(&self, window: &(impl gtk::prelude::IsA<gtk::Window> + gtk4::prelude::DragExt)) -> bool {
    |                                          +                                    +++++++++++++++++++++++++
334 |     pub fn set_parent_of(&self, window: &(impl gtk::prelude::IsA<gtk::Window> + gtk4::prelude::DrawContextExt)) -> bool {
    |                                          +                                    ++++++++++++++++++++++++++++++++
334 |     pub fn set_parent_of(&self, window: &(impl gtk::prelude::IsA<gtk::Window> + gtk4::prelude::GLContextExt)) -> bool {
    |                                          +                                    ++++++++++++++++++++++++++++++
334 |     pub fn set_parent_of(&self, window: &(impl gtk::prelude::IsA<gtk::Window> + gtk4::prelude::GskRendererExt)) -> bool {
    |                                          +                                    ++++++++++++++++++++++++++++++++
      and 1 other candidate

error[E0277]: the trait bound `impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<gtk4::Widget>` is not satisfied
    --> src/window_identifier/mod.rs:340:36
     |
340  |                 WidgetExt::realize(window);
     |                 ------------------ ^^^^^^ the trait `glib::object::IsA<gtk4::Widget>` is not implemented for `impl gtk::prelude::IsA<gtk::Window>`
     |                 |
     |                 required by a bound introduced by this call
     |
note: required by a bound in `gtk4::prelude::WidgetExt::realize`
    --> /home/arun-mani-j/.config/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/gtk4-0.9.5/src/auto/widget.rs:58:22
     |
58   | pub trait WidgetExt: IsA<Widget> + sealed::Sealed + 'static {
     |                      ^^^^^^^^^^^ required by this bound in `WidgetExt::realize`
...
1077 |     fn realize(&self) {
     |        ------- required by a bound in this associated function

error[E0599]: the method `surface` exists for reference `&impl gtk::prelude::IsA<gtk::Window>`, but its trait bounds were not satisfied
   --> src/window_identifier/mod.rs:341:24
    |
341 |                 window.surface().unwrap()
    |                        ^^^^^^^ method cannot be called on `&impl gtk::prelude::IsA<gtk::Window>` due to unsatisfied trait bounds
    |
    = note: the following trait bounds were not satisfied:
            `impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<Drag>`
            which is required by `impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::DragExt`
            `impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<DrawContext>`
            which is required by `impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::DrawContextExt`
            `impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<GLContext>`
            which is required by `impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::GLContextExt`
            `impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<gtk4::gsk4::Renderer>`
            which is required by `impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::GskRendererExt`
            `impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<Native>`
            which is required by `impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::NativeExt`
            `&impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<Drag>`
            which is required by `&impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::DragExt`
            `&impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<DrawContext>`
            which is required by `&impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::DrawContextExt`
            `&impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<GLContext>`
            which is required by `&impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::GLContextExt`
            `&impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<gtk4::gsk4::Renderer>`
            which is required by `&impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::GskRendererExt`
            `&impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<Native>`
            which is required by `&impl gtk::prelude::IsA<gtk::Window>: gtk4::prelude::NativeExt`
    = help: items from traits can only be used if the type parameter is bounded by the trait
help: the following traits define an item `surface`, perhaps you need to restrict type parameter `impl gtk::prelude::IsA<gtk::Window>` with one of them:
    |
334 |     pub fn set_parent_of(&self, window: &(impl gtk::prelude::IsA<gtk::Window> + gtk4::prelude::DragExt)) -> bool {
    |                                          +                                    +++++++++++++++++++++++++
334 |     pub fn set_parent_of(&self, window: &(impl gtk::prelude::IsA<gtk::Window> + gtk4::prelude::DrawContextExt)) -> bool {
    |                                          +                                    ++++++++++++++++++++++++++++++++
334 |     pub fn set_parent_of(&self, window: &(impl gtk::prelude::IsA<gtk::Window> + gtk4::prelude::GLContextExt)) -> bool {
    |                                          +                                    ++++++++++++++++++++++++++++++
334 |     pub fn set_parent_of(&self, window: &(impl gtk::prelude::IsA<gtk::Window> + gtk4::prelude::GskRendererExt)) -> bool {
    |                                          +                                    ++++++++++++++++++++++++++++++++
      and 1 other candidate

error[E0277]: the trait bound `impl gtk::prelude::IsA<gtk::Window>: glib::object::IsA<gtk4::Widget>` is not satisfied
   --> src/window_identifier/mod.rs:353:56
    |
353 |                 let display = match WidgetExt::display(window).dynamic_cast::<X11Display>() {
    |                                     ------------------ ^^^^^^ the trait `glib::object::IsA<gtk4::Widget>` is not implemented for `impl gtk::prelude::IsA<gtk::Window>`
    |                                     |
    |                                     required by a bound introduced by this call
    |
note: required by a bound in `gtk4::prelude::WidgetExt::display`
   --> /home/arun-mani-j/.config/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/gtk4-0.9.5/src/auto/widget.rs:58:22
    |
58  | pub trait WidgetExt: IsA<Widget> + sealed::Sealed + 'static {
    |                      ^^^^^^^^^^^ required by this bound in `WidgetExt::display`
...
425 |     fn display(&self) -> gdk::Display {
    |        ------- required by a bound in this associated function

Some errors have detailed explanations: E0277, E0599.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `ashpd` (lib) due to 4 previous errors
```

bilelmoussaoui replied:
you can inside the function do `let window = window.as_ref();` because IsA is just there for constraints, the conversion is done through AsRef

arun-mani-j replied:
Oh, did not know, thanks, done.

bilelmoussaoui commented on Cargo.toml:
```diff
@@ -33,7 +33,7 @@ enumflags2 = "0.7"
 futures-util = "0.3"
 gdk4wayland = { package = "gdk4-wayland", version = "0.9", optional = true }
-gdk4x11 = { package = "gdk4-x11", version = "0.9", optional = true }
+gdk4x11 = { package = "gdk4-x11", version = "0.9", optional = true, features = [ "xlib" ] }
```
this will enable the xlib feature all the time, whether you enable the backend feature or not... not sure what to do here

arun-mani-j replied:
I think it's fine as there won't be any breaking change in `xlib` and so we need not worry about breaking API contract?

bilelmoussaoui replied:
as the dependency itself is rather minimal, let us not block the situation forever.

arun-mani-j commented on src/window_identifier/mod.rs:
```diff
@@ -310,6 +315,94 @@ impl Serialize for WindowIdentifierType {
+    /// Might realize the window if it is not realized yet.
+    ///
+    /// Returns `true` on success.
+    #[cfg(all(
+        any(feature = "gtk4_wayland", feature = "gtk4_x11"),
+        feature = "backend"
+    ))]
```
Thanks, done.

bilelmoussaoui commented on src/window_identifier/mod.rs:
```diff
@@ -310,6 +315,109 @@ impl Serialize for WindowIdentifierType {
 
+impl WindowIdentifierType {
+    /// Sets the given window as a modal child of the window represented by
+    /// self. Currently, setting Wayland window as parent of Wayland window and
+    /// X11 window as parent of X11 window is supported.
```
Can you instead write this as a list of cases and mention next to each one whether it is a supported one or not?

arun-mani-j replied:
Done.

bilelmoussaoui commented:
awesome, thank you
